haddock-library/src/Documentation/Haddock/Parser/Util.hs

```diff
@@ -13,60 +13,73 @@
 , takeHorizontalSpace
 , skipHorizontalSpace
 ) where
 
+import qualified Data.Text as T
+import           Data.Text (Text)
+
 import           Control.Applicative
 import           Control.Monad (mfilter)
-import           Documentation.Haddock.Parser.Monad hiding (isHorizontalSpace)
-import           Data.ByteString.Char8 (ByteString)
-import qualified Data.ByteString.Char8 as BS
+import           Documentation.Haddock.Parser.Monad
 import           Prelude hiding (takeWhile)
 
-#if MIN_VERSION_bytestring(0,10,2)
-import           Data.ByteString.Char8 (unsnoc)
-#else
-unsnoc :: ByteString -> Maybe (ByteString, Char)
+import           Data.Maybe (isJust)
+import           Data.Char (isSpace)
+
+-- | Like 'T.uncons', but for the last character instead of the first.
+unsnoc :: Text -> Maybe (Text, Char)
 unsnoc bs
-  | BS.null bs = Nothing
-  | otherwise = Just (BS.init bs, BS.last bs)
-#endif
+  | T.null bs = Nothing
+  | otherwise = Just (T.init bs, T.last bs)
 
 -- | Remove all leading and trailing whitespace
-strip :: String -> String
-strip = (\f -> f . f) $ dropWhile isSpace . reverse
+strip :: Text -> Text  
+strip = T.strip 
+
+-- | Like 'elem' but for 'Text'
+inClass :: Char -> Text -> Bool
+inClass c = isJust . T.find (== c)
 
-isHorizontalSpace :: Char -> Bool
-isHorizontalSpace = inClass " \t\f\v\r"
+-- | Like 'notElem' but for 'Text'
+notInClass :: Char -> Text -> Bool
+notInClass c = not . inClass c 
 
+-- | Characters that count as horizontal space
+horizontalSpace :: Text
+horizontalSpace = " \t\f\v\r"
+
+-- | Skip and ignore leading horizontal space
 skipHorizontalSpace :: Parser ()
-skipHorizontalSpace = skipWhile isHorizontalSpace
+skipHorizontalSpace = skipWhile (`inClass` horizontalSpace)
 
-takeHorizontalSpace :: Parser BS.ByteString
-takeHorizontalSpace = takeWhile isHorizontalSpace
+-- | Take leading horizontal space
+takeHorizontalSpace :: Parser Text 
+takeHorizontalSpace = takeWhile (`inClass` horizontalSpace)
 
-makeLabeled :: (String -> Maybe String -> a) -> String -> a
-makeLabeled f input = case break isSpace $ removeEscapes $ strip input of
-  (uri, "")    -> f uri Nothing
-  (uri, label) -> f uri (Just $ dropWhile isSpace label)
+makeLabeled :: (String -> Maybe String -> a) -> Text -> a
+makeLabeled f input = case T.break isSpace $ removeEscapes $ strip input of
+  (uri, "")    -> f (T.unpack uri) Nothing
+  (uri, label) -> f (T.unpack uri) (Just . T.unpack $ T.stripStart label)
 
 -- | Remove escapes from given string.
 --
 -- Only do this if you do not process (read: parse) the input any further.
-removeEscapes :: String -> String
-removeEscapes "" = ""
-removeEscapes ('\\':'\\':xs) = '\\' : removeEscapes xs
-removeEscapes ('\\':xs) = removeEscapes xs
-removeEscapes (x:xs) = x : removeEscapes xs
-
-takeUntil :: ByteString -> Parser ByteString
-takeUntil end_ = dropEnd <$> requireEnd (scan (False, end) p) >>= gotSome
+removeEscapes :: Text -> Text
+removeEscapes = T.filter (/= '\\') . T.replace "\\\\" "\\"
```

Comment:
Didn't you change semantics? The previous version would leave a single slash where two were present while this first replaces double slashes with single ones and then removes all slashes anyway.I thought we had a test for this...